src/Hint/Restrict.hs

```diff
@@ -190,30 +196,31 @@ checkImports modu lImportDecls (def, mp) = mapMaybe getImportHint lImportDecls
               case fromMaybe ImportStyleUnrestricted $ getAlt riImportStyle of
                 ImportStyleUnrestricted
                   | NotQualified <- ideclQualified -> (Nothing, Nothing)
-                  | otherwise -> (second (<> " or unqualified") <$> expectedQualStyle, Nothing)
-                ImportStyleQualified -> (expectedQualStyleDef, Nothing)
+                  | otherwise -> (Just $ second (<> " or unqualified") expectedQualStyle, Nothing)
+                ImportStyleQualified -> (Just expectedQualStyle, Nothing)
                 ImportStyleExplicitOrQualified
                   | Just (False, _) <- first (== EverythingBut) <$> ideclImportList -> (Nothing, Nothing)
                   | otherwise ->
-                      ( second (<> " or with an explicit import list") <$> expectedQualStyleDef
+                      ( Just $ second (<> " or with an explicit import list") expectedQualStyle
                       , Nothing )
                 ImportStyleExplicit
                   | Just (False, _) <- first (== EverythingBut) <$> ideclImportList -> (Nothing, Nothing)
                   | otherwise ->
-                      ( Just (NotQualified, "unqualified")
+                      ( Just (Right NotQualified, "unqualified")
                       , Just $ Just (Exactly, noLocA []) )
-                ImportStyleUnqualified -> (Just (NotQualified, "unqualified"), Nothing)
-            expectedQualStyleDef = expectedQualStyle <|> Just (QualifiedPre, "qualified")
+                ImportStyleUnqualified -> (Just (Right NotQualified, "unqualified"), Nothing)
             expectedQualStyle =
               case fromMaybe QualifiedStyleUnrestricted $ getAlt riQualifiedStyle of
-                QualifiedStyleUnrestricted -> Nothing
-                QualifiedStylePost -> Just (QualifiedPost, "post-qualified")
-                QualifiedStylePre -> Just (QualifiedPre, "pre-qualified")
+                QualifiedStyleUnrestricted -> (Left QualifiedPostOrPre, "qualified")
+                QualifiedStylePost -> (Right QualifiedPost, "post-qualified")
+                QualifiedStylePre -> (Right QualifiedPre, "pre-qualified")
             qualIdea
-              | Just ideclQualified == (fst <$> expectedQual) = Nothing
+              | Just (Right ideclQualified) == (fst <$> expectedQual) = Nothing
```

Comment:
The logic of this function seems quite convoluted. Can you add some comments in this function, for example, what does it mean when `expectedQual` is a `Just` or a `Nothing`? `expectedQual` doesn't seem to represent what its name suggests - the expected qualifying style, because the code for obtaining `expectedQual` also checks the actual qualifying style from `ideclQualified`.

Reply:
Indeed.

I tried to go with a minimal fix, rather then trying to rework the whole function. Focused patches are typically easier to review.

`expectedQual` gets determined by this beefy case-match (which is easy to see, just lexically):

https://github.com/ndmitchell/hlint/blob/0343ec5d68667ce21188cb7871e7ec0d5ac79671/src/Hint/Restrict.hs#L195-L211

It handles the rule's variations in `importStyle` (together with nuances around `asRequired`, `qualifiedStyle`, hiding lists, etc). It also looks into the declaration-under-check indeed, the `ideclQualified` — AFAICS, just to spare having a second case-match over `riImportStyle`.

TL;DR: if `expectedQual` is Just something, it will emit a hint... unless silenced by the 2 checks in `qualIdea`.

I'll document this in-source.